core/src/main/java/org/apache/carbondata/core/datamap/status/DataMapStatusManager.java

```diff
@@ -127,4 +150,115 @@ private static DataMapSchema getDataMapSchema(String dataMapName)
   }
 
+  /**
+   * Returns valid segment list for a given RelationIdentifier
+   *
+   * @param relationIdentifier
+   * @return
+   * @throws IOException
+   */
+  public static List<String> getSegmentList(RelationIdentifier relationIdentifier)
+      throws IOException {
+    List<String> segmentList = new ArrayList<>();
+    AbsoluteTableIdentifier absoluteTableIdentifier =
+        AbsoluteTableIdentifier.from(relationIdentifier.getTablePath());
+    List<Segment> validSegments =
+        new SegmentStatusManager(absoluteTableIdentifier).getValidAndInvalidSegments()
+            .getValidSegments();
+    for (Segment segment : validSegments) {
+      segmentList.add(segment.getSegmentNo());
+    }
+    return segmentList;
+  }
+
+  /**
+   * This method will delete segment folders of the mv datamap table and update the
+   * datamapSegmentStatus map in case of Insert-Overwrite/Update operation on main table
+   */
+  public static void cleanMVdatamap(CarbonTable carbonTable) throws IOException {
+    List<DataMapSchema> allDataMapSchemas =
+        DataMapStoreManager.getInstance().getDataMapSchemasOfTable(carbonTable);
+    for (DataMapSchema datamapschema : allDataMapSchemas) {
+      if (datamapschema.getProviderName()
+          .equalsIgnoreCase(DataMapClassProvider.MV.getShortName())) {
+        CarbonTable datamapTable = CarbonTable
+            .buildFromTablePath(datamapschema.getRelationIdentifier().getTableName(),
+                datamapschema.getRelationIdentifier().getDatabaseName(),
+                datamapschema.getRelationIdentifier().getTablePath(),
+                datamapschema.getRelationIdentifier().getTableId());
+        SegmentStatusManager segmentStatusManager =
+            new SegmentStatusManager(datamapTable.getAbsoluteTableIdentifier());
+        ICarbonLock carbonLock = segmentStatusManager.getTableStatusLock();
+        try {
+          if (carbonLock.lockWithRetries()) {
+            LOGGER.info(
+                "Acquired lock for table" + datamapTable.getDatabaseName() + "." + datamapTable
+                    .getTableName() + " for table status updation");
+            LoadMetadataDetails[] loadMetadataDetails =
+                SegmentStatusManager.readLoadMetadata(datamapTable.getMetadataPath());
+            List<CarbonFile> staleFolders = new ArrayList<>();
+            for (LoadMetadataDetails entry : loadMetadataDetails) {
+              entry.setSegmentStatus(SegmentStatus.MARKED_FOR_DELETE);
+              // For insert overwrite, we will delete the old segment folder immediately
+              // So collect the old segments here
+              String segmentPath = CarbonTablePath
+                  .getSegmentPath(datamapschema.getRelationIdentifier().getTablePath(),
+                      entry.getLoadName());
+              if (FileFactory.isFileExist(segmentPath, FileFactory.getFileType(segmentPath))) {
+                staleFolders.add(FileFactory.getCarbonFile(segmentPath));
```

Comment:
Please don't remove immediately as some queries might be running on it. Let it remove during next time load